askama/src/filters/indent.rs

```diff
@@ -0,0 +1,355 @@
+/// to indent the first and blank lines, resp.
+///
+/// ### Example of `indent` with spaces
+///
+/// ```
+/// # #[cfg(feature = "code-in-doc")] {
+/// # use askama::Template;
+/// /// ```jinja
+/// /// <div>{{ example|indent(4) }}</div>
+/// /// ```
+/// #[derive(Template)]
+/// #[template(ext = "html", in_doc = true)]
+/// struct Example<'a> {
+///     example: &'a str,
+/// }
+///
+/// assert_eq!(
+///     Example { example: "hello\nfoo\nbar" }.to_string(),
+///     "<div>hello\n    foo\n    bar</div>"
+/// );
+/// # }
+/// ```
+///
+/// ### Example of `indent` with prefix a custom prefix
+///
+/// ```
+/// # #[cfg(feature = "code-in-doc")] {
+/// # use askama::Template;
+/// /// ```jinja
+/// /// <div>{{ example|indent("$$$ ") }}</div>
+/// /// ```
+/// #[derive(Template)]
+/// #[template(ext = "html", in_doc = true)]
+/// struct Example<'a> {
+///     example: &'a str,
+/// }
+///
+/// assert_eq!(
+///     Example { example: "hello\nfoo\nbar" }.to_string(),
+///     "<div>hello\n$$$ foo\n$$$ bar</div>"
+/// );
+/// # }
+/// ```
+#[inline]
+pub fn indent<S, I: AsIndent>(
+    source: S,
+    indent: I,
+    first: bool,
+    blank: bool,
+) -> Result<Indent<S, I>, Infallible> {
+    Ok(Indent {
+        source,
+        indent,
+        first,
+        blank,
+    })
+}
+
+pub struct Indent<S, I> {
+    source: S,
+    indent: I,
+    first: bool,
+    blank: bool,
+}
+
+impl<S: fmt::Display, I: AsIndent> fmt::Display for Indent<S, I> {
+    fn fmt(&self, dest: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let indent = self.indent.as_indent();
+        write!(
+            IndentWriter::new(dest, indent, self.first, self.blank),
+            "{}",
+            self.source
+        )?;
+        Ok(())
+    }
+}
+
+impl<S: FastWritable, I: AsIndent> FastWritable for Indent<S, I> {
+    fn write_into<W: fmt::Write + ?Sized>(
+        &self,
+        dest: &mut W,
+        values: &dyn crate::Values,
+    ) -> crate::Result<()> {
+        let indent = self.indent.as_indent();
+        self.source.write_into(
+            &mut IndentWriter::new(dest, indent, self.first, self.blank),
+            values,
+        )?;
+        Ok(())
+    }
+}
+
+struct IndentWriter<'a, W> {
+    dest: W,
+    indent: &'a str,
+    first: bool,
+    blank: bool,
+    is_new_line: bool,
+}
+
+impl<'a, W: fmt::Write> IndentWriter<'a, W> {
+    fn new(dest: W, indent: &'a str, first: bool, blank: bool) -> Self {
+        IndentWriter {
+            dest,
+            indent,
+            first,
+            blank,
+            is_new_line: true,
+        }
+    }
+}
+
+impl<W: fmt::Write> fmt::Write for IndentWriter<'_, W> {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        if self.indent.is_empty() {
+            return self.dest.write_str(s);
+        }
+
+        for (idx, line) in s.split_inclusive('\n').enumerate() {
+            if (self.first || idx > 0 || !self.is_new_line)
```

Comment:
There is only one first line :), so `idx` would need to be tracked across calls. But in here we can simply removed the counter and set `self.first = true` at the end of the loop.